src/System.Numerics.Tensors/System/Numerics/CompressedSparseTensor.cs

```diff
@@ -342,7 +349,63 @@ public override Tensor<T> Reshape(params int[] dimensions)
                 newCompressedCounts[newCompressedIndex + 1] = valueIndex + 1;
             }
 
-            return new CompressedSparseTensor<T>(newValues, newCompressedCounts, newIndices, valueCount, dimensions, IsReversedStride);
+            return new CompressedSparseTensor<T>(newValues, newCompressedCounts, newIndices, nonZeroCount, dimensions, IsReversedStride);
+        }
+
+        public override DenseTensor<T> ToDenseTensor()
+        {
+            var denseTensor = new DenseTensor<T>(dimensions, reverseStride: IsReversedStride);
+
+            var compressedIndex = 0;
+
+            for (int valueIndex = 0; valueIndex < nonZeroCount; valueIndex++)
+            {
+                while (valueIndex >= compressedCounts[compressedIndex + 1])
+                {
+                    compressedIndex++;
+                    Debug.Assert(compressedIndex < compressedCounts.Length);
+                }
+
+                var index = indices[valueIndex] + compressedIndex * strides[compressedDimension];
+
+                denseTensor.Buffer[index] = values[valueIndex];
+            }
+
+            return denseTensor;
+        }
+
+        public override CompressedSparseTensor<T> ToCompressedSparseTensor()
+        {
+            var newValues = new T[NonZeroCount];
+            Array.Copy(values, newValues, NonZeroCount);
+
+            var newIndicies = new int[NonZeroCount];
+            Array.Copy(indices, newIndicies, NonZeroCount);
+
+            return new CompressedSparseTensor<T>(newValues, (int[])compressedCounts.Clone(), newIndicies, nonZeroCount, dimensions, IsReversedStride);
+        }
+
+        public override SparseTensor<T> ToSparseTensor()
+        {
+            var sparseTensor = new SparseTensor<T>(dimensions, capacity: NonZeroCount, reverseStride: IsReversedStride);
+
+            var compressedIndex = 0;
+
+            for (int valueIndex = 0; valueIndex < nonZeroCount; valueIndex++)
```

Comment:
(minor) Not sure if this algorithm would ever change, but there is some duplicated logic between this and `ToDenseTensor` above.  Would it make sense to refactor this logic in a single spot?

Reply:
I would hope so, once we get https://github.com/dotnet/corefxlab/issues/1763.  

For something as small as this I wouldn't try and factor it out since the body of the loop is subtly different and I don't want to inject a delegate for that difference.